File: src/lib/pdfGenerator.ts
import { jsPDF } from "jspdf";
import autoTable from 'jspdf-autotable';
import { Database } from '../types/supabase';
import { format } from 'date-fns';
import { es } from 'date-fns/locale';

type Event = Database['public']['Tables']['events']['Row'] & {
  clients?: Database['public']['Tables']['clients']['Row'] | null;
};
type Profile = Database['public']['Tables']['users']['Row'];
type ProductItem = Database['public']['Tables']['event_products']['Row'] & {
  products: { name: string } | null;
};
type ExtraItem = Database['public']['Tables']['event_extras']['Row'];

const BRAND_COLOR = '#FF6B35'; // Naranja de la marca
const TEXT_COLOR = '#333333';
const GRAY_COLOR = '#666666';

// Helper para formatear moneda
const formatCurrency = (amount: number) => {
  return new Intl.NumberFormat('es-MX', {
    style: 'currency',
    currency: 'MXN',
  }).format(amount);
};

// Helper para agregar encabezado común
const addHeader = (doc: jsPDF, profile: Profile | null, title: string): number => {
  const pageWidth = doc.internal.pageSize.width;
  
  // Business Name
  doc.setFontSize(20);
  doc.setTextColor(BRAND_COLOR);
  doc.text(profile?.business_name || profile?.name || 'EventosApp', 20, 20);
  
  // Title
  doc.setFontSize(24);
  doc.setTextColor(TEXT_COLOR);
  doc.text(title.toUpperCase(), pageWidth - 20, 20, { align: 'right' });
  
  // Line separator
  doc.setDrawColor(BRAND_COLOR);
  doc.setLineWidth(1);
  doc.line(20, 25, pageWidth - 20, 25);
  
  return 35; // Returns Y position for next content
};

export const generateBudgetPDF = (
  event: Event,
  profile: Profile | null,
  products: ProductItem[],
  extras: ExtraItem[]
) => {
  const doc = new jsPDF();
  let currentY = addHeader(doc, profile, 'Presupuesto');
  
  // Info del Cliente y Evento
  doc.setFontSize(10);
  doc.setTextColor(GRAY_COLOR);
  
  // Columna Izquierda (Cliente)
  doc.setFont('helvetica', 'bold');
  doc.text('INFORMACIÓN DEL CLIENTE', 20, currentY);
  doc.setFont('helvetica', 'normal');
  doc.text(event.clients?.name || 'Cliente no registrado', 20, currentY + 5);
  if (event.clients?.phone) doc.text(`Tel: ${event.clients.phone}`, 20, currentY + 10);
  if (event.clients?.email) doc.text(`Email: ${event.clients.email}`, 20, currentY + 15);
  
  // Columna Derecha (Evento)
  const rightColX = doc.internal.pageSize.width / 2 + 10;
  doc.setFont('helvetica', 'bold');
  doc.text('DETALLES DEL EVENTO', rightColX, currentY);
  doc.setFont('helvetica', 'normal');
  const eventDate = new Date(event.event_date);
  // Ajuste de zona horaria simple para visualización (evitar desfase por UTC)
  const userTimezoneOffset = eventDate.getTimezoneOffset() * 60000;
  const localDate = new Date(eventDate.getTime() + userTimezoneOffset);

  doc.text(`Fecha: ${format(localDate, "d 'de' MMMM, yyyy", { locale: es })}`, rightColX, currentY + 5);
  doc.text(`Tipo: ${event.service_type}`, rightColX, currentY + 10);
  doc.text(`Personas: ${event.num_people}`, rightColX, currentY + 15);
  
  currentY += 25;

  // Tabla de Productos
  const productRows = products.map((p) => [
    p.products?.name || 'Producto',
    p.quantity.toString(),
    formatCurrency(p.unit_price),
    formatCurrency((p.unit_price - (p.discount || 0)) * p.quantity)
  ]);
  
  // Tabla de Extras
  const extraRows = extras.map((e) => [
    e.description,
    "1",
    formatCurrency(e.price),
    formatCurrency(e.price)
  ]);
  
  const body = [...productRows, ...extraRows];
  
  if (body.length > 0) {
    autoTable(doc, {
      startY: currentY,
      head: [['Descripción', 'Cant.', 'Precio Unit.', 'Total']],
      body: body,
      headStyles: { fillColor: BRAND_COLOR },
      styles: { fontSize: 9 },
      theme: 'grid',
    });
    
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    currentY = (doc as any).lastAutoTable.finalY + 10;
  } else {
    currentY += 10;
    doc.text('No hay productos o servicios registrados.', 20, currentY);
    currentY += 10;
  }
  
  // Totales
  const pageWidth = doc.internal.pageSize.width;
  const rightMargin = pageWidth - 20;
  
  doc.setFontSize(10);
  doc.setTextColor(TEXT_COLOR);
  
  const totalRevenue = event.total_amount || 0;
  const taxAmount = event.tax_amount || 0;
  // Si requiere factura, el total incluye IVA, así que el subtotal es menos el IVA.
  // Si no requiere, el total es neto.
  const subtotal = event.requires_invoice ? totalRevenue - taxAmount : totalRevenue;

  doc.text(`Subtotal: ${formatCurrency(subtotal)}`, rightMargin, currentY, { align: 'right' });
  
  if (event.requires_invoice) {
    currentY += 5;
    doc.text(`IVA (${event.tax_rate || 16}%): ${formatCurrency(taxAmount)}`, rightMargin, currentY, { align: 'right' });
  }
  
  currentY += 7;
  doc.setFontSize(12);
  doc.setFont('helvetica', 'bold');
  doc.text(`TOTAL: ${formatCurrency(totalRevenue)}`, rightMargin, currentY, { align: 'right' });
  
  // Pie de página
  const pageHeight = doc.internal.pageSize.height;
  doc.setFontSize(8);
  doc.setFont('helvetica', 'normal');
  doc.setTextColor(GRAY_COLOR);
  doc.text('Este presupuesto tiene una validez de 15 días.', 20, pageHeight - 10);
  
  doc.save(`Presupuesto_${event.clients?.name || 'Cliente'}.pdf`);
};

export const generateContractPDF = (
  event: Event,
  profile: Profile | null
) => {
  const doc = new jsPDF();
  const currentY = addHeader(doc, profile, 'Contrato');
  
  doc.setFontSize(10);
  doc.setTextColor(TEXT_COLOR);
  
  const city = event.city || profile?.business_name || 'la ciudad';
  const dateStr = format(new Date(), "d 'de' MMMM 'de' yyyy", { locale: es });
  const providerName = profile?.business_name || profile?.name || 'EL PROVEEDOR';
  const clientName = event.clients?.name || 'EL CLIENTE';
  
  const eventDate = new Date(event.event_date);
  const userTimezoneOffset = eventDate.getTimezoneOffset() * 60000;
  const localDate = new Date(eventDate.getTime() + userTimezoneOffset);
  const eventDateStr = format(localDate, "d 'de' MMMM 'de' yyyy", { locale: es });

  const text = `En ${city}, a los ${dateStr}, comparecen por una parte ${providerName} (en adelante "EL PROVEEDOR"), y por la otra parte ${clientName} (en adelante "EL CLIENTE"), quienes convienen en celebrar el presente Contrato de Prestación de Servicios.

PRIMERA: OBJETO.
EL PROVEEDOR se compromete a prestar los servicios de ${event.service_type} para el evento que se llevará a cabo el día ${eventDateStr}.

SEGUNDA: DETALLES DEL EVENTO.
El evento contará con aproximadamente ${event.num_people} personas.
Ubicación: ${event.location || 'A definir'}.
Horario: ${event.start_time || 'A definir'} - ${event.end_time || 'A definir'}.

TERCERA: COSTO Y FORMA DE PAGO.
El costo total del servicio es de ${formatCurrency(event.total_amount)}.
EL CLIENTE realizará un anticipo del ${event.deposit_percent ?? 50}% para reservar la fecha, debiendo liquidar el saldo restante antes del inicio del evento.

CUARTA: CANCELACIONES.
En caso de cancelación por parte de EL CLIENTE con menos de ${event.cancellation_days ?? 15} días de anticipación, el anticipo no será reembolsado.
${event.refund_percent && event.refund_percent > 0 ? `Si la cancelación se realiza con la debida anticipación, se reembolsará el ${event.refund_percent}% del anticipo entregado.` : ''}

QUINTA: ACEPTACIÓN.
Ambas partes aceptan los términos y condiciones del presente contrato.`;

  const splitText = doc.splitTextToSize(text, 170);
  doc.text(splitText, 20, currentY + 10);
  
  // Firmas
  const pageHeight = doc.internal.pageSize.height;
  const signY = pageHeight - 40;
  
  doc.line(20, signY, 80, signY);
  doc.text('EL PROVEEDOR', 50, signY + 5, { align: 'center' });
  doc.text(providerName, 50, signY + 10, { align: 'center', maxWidth: 60 });
  
  doc.line(130, signY, 190, signY);
  doc.text('EL CLIENTE', 160, signY + 5, { align: 'center' });
  doc.text(clientName, 160, signY + 10, { align: 'center', maxWidth: 60 });
  
  doc.save(`Contrato_${event.clients?.name || 'Cliente'}.pdf`);
};

File: flutter/lib/core/utils/pdf_generator.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:path_provider/path_provider.dart';
import 'package:share_plus/share_plus.dart';
import 'package:syncfusion_flutter_pdf/pdf.dart';

import 'package:eventosapp/features/events/domain/entities/event_entity.dart';
import 'package:eventosapp/features/events/domain/entities/event_product_entity.dart';
import 'package:eventosapp/features/events/domain/entities/event_extra_entity.dart';

// Brand color naranja
const _brandR = 255;
const _brandG = 107;
const _brandB = 53;

String _formatCurrency(double amount) {
  return NumberFormat.currency(locale: 'es_MX', symbol: r'$').format(amount);
}

String _formatDate(DateTime date) {
  return DateFormat("d 'de' MMMM 'de' yyyy", 'es').format(date);
}

// ─────────────────────────────────────────────
// PRESUPUESTO PDF
// ─────────────────────────────────────────────
Future<void> generateBudgetPDF({
  required EventEntity event,
  required List<EventProductEntity> products,
  required List<EventExtraEntity> extras,
  String businessName = 'EventosApp',
}) async {
  final document = PdfDocument();
  final page = document.pages.add();
  final graphics = page.graphics;
  final pageWidth = page.getClientSize().width;

  final brandColor = PdfColor(_brandR, _brandG, _brandB);
  final grayColor = PdfColor(102, 102, 102);
  final textColor = PdfColor(51, 51, 51);

  final boldFont =
      PdfStandardFont(PdfFontFamily.helvetica, 10, style: PdfFontStyle.bold);
  final regularFont = PdfStandardFont(PdfFontFamily.helvetica, 9);
  final titleFont =
      PdfStandardFont(PdfFontFamily.helvetica, 20, style: PdfFontStyle.bold);
  final headerFont =
      PdfStandardFont(PdfFontFamily.helvetica, 24, style: PdfFontStyle.bold);

  double y = 20;

  // ── Encabezado ──
  graphics.drawString(
    businessName,
    titleFont,
    brush: PdfSolidBrush(brandColor),
    bounds: Rect.fromLTWH(20, y, pageWidth / 2, 25),
  );
  graphics.drawString(
    'PRESUPUESTO',
    headerFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(0, y, pageWidth - 20, 25),
    format: PdfStringFormat(alignment: PdfTextAlignment.right),
  );
  y += 30;

  // Línea separadora
  graphics.drawLine(
    PdfPen(brandColor, width: 1.5),
    Offset(20, y),
    Offset(pageWidth - 20, y),
  );
  y += 12;

  // ── Info cliente & evento (2 columnas) ──
  final halfW = (pageWidth - 40) / 2;
  final rightX = 20 + halfW + 10;

  graphics.drawString('INFORMACIÓN DEL CLIENTE', boldFont,
      brush: PdfSolidBrush(grayColor), bounds: Rect.fromLTWH(20, y, halfW, 12));
  graphics.drawString('DETALLES DEL EVENTO', boldFont,
      brush: PdfSolidBrush(grayColor),
      bounds: Rect.fromLTWH(rightX, y, halfW, 12));
  y += 14;

  graphics.drawString(event.clientName, regularFont,
      brush: PdfSolidBrush(textColor), bounds: Rect.fromLTWH(20, y, halfW, 12));
  graphics.drawString(
    'Fecha: ${_formatDate(event.eventDate)}',
    regularFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(rightX, y, halfW, 12),
  );
  y += 12;

  graphics.drawString(
    'Tipo: ${event.serviceType}',
    regularFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(rightX, y, halfW, 12),
  );
  y += 12;

  graphics.drawString(
    'Personas: ${event.numPeople}',
    regularFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(rightX, y, halfW, 12),
  );
  y += 20;

  // ── Tabla de productos y extras ──
  final grid = PdfGrid();
  grid.columns.add(count: 4);
  grid.columns[0].width = pageWidth * 0.45;
  grid.columns[1].width = pageWidth * 0.10;
  grid.columns[2].width = pageWidth * 0.20;
  grid.columns[3].width = pageWidth * 0.25;

  final headerRow = grid.headers.add(1)[0];
  headerRow.cells[0].value = 'Descripción';
  headerRow.cells[1].value = 'Cant.';
  headerRow.cells[2].value = 'Precio Unit.';
  headerRow.cells[3].value = 'Total';

  final headerStyle = PdfGridCellStyle(
    backgroundBrush: PdfSolidBrush(brandColor),
    textBrush: PdfSolidBrush(PdfColor(255, 255, 255)),
    font: boldFont,
  );
  for (int i = 0; i < 4; i++) {
    headerRow.cells[i].style = headerStyle;
    headerRow.cells[i].stringFormat =
        PdfStringFormat(alignment: PdfTextAlignment.center);
  }

  for (final p in products) {
    final row = grid.rows.add();
    row.cells[0].value = p.productName;
    row.cells[1].value = p.quantity.toStringAsFixed(0);
    row.cells[2].value = _formatCurrency(p.unitPrice);
    row.cells[3].value =
        _formatCurrency((p.unitPrice - p.discount) * p.quantity);
  }

  for (final e in extras) {
    final row = grid.rows.add();
    row.cells[0].value = e.description;
    row.cells[1].value = '1';
    row.cells[2].value = _formatCurrency(e.price);
    row.cells[3].value = _formatCurrency(e.price);
  }

  grid.style = PdfGridStyle(
    font: regularFont,
    cellPadding: PdfPaddings(left: 5, right: 5, top: 3, bottom: 3),
  );

  final result = grid.draw(
    page: page,
    bounds: Rect.fromLTWH(20, y, pageWidth - 40, 0),
  );
  y = result!.bounds.bottom + 12;

  // ── Totales ──
  final subtotal = event.requiresInvoice
      ? event.totalAmount - event.taxAmount
      : event.totalAmount;

  graphics.drawString(
    'Subtotal: ${_formatCurrency(subtotal)}',
    regularFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(0, y, pageWidth - 20, 12),
    format: PdfStringFormat(alignment: PdfTextAlignment.right),
  );
  y += 14;

  if (event.requiresInvoice) {
    graphics.drawString(
      'IVA (${event.taxRate.toStringAsFixed(0)}%): ${_formatCurrency(event.taxAmount)}',
      regularFont,
      brush: PdfSolidBrush(textColor),
      bounds: Rect.fromLTWH(0, y, pageWidth - 20, 12),
      format: PdfStringFormat(alignment: PdfTextAlignment.right),
    );
    y += 14;
  }

  graphics.drawString(
    'TOTAL: ${_formatCurrency(event.totalAmount)}',
    boldFont,
    brush: PdfSolidBrush(brandColor),
    bounds: Rect.fromLTWH(0, y, pageWidth - 20, 14),
    format: PdfStringFormat(alignment: PdfTextAlignment.right),
  );

  // Pie
  final pageH = page.getClientSize().height;
  graphics.drawString(
    'Este presupuesto tiene una validez de 15 días.',
    PdfStandardFont(PdfFontFamily.helvetica, 8),
    brush: PdfSolidBrush(grayColor),
    bounds: Rect.fromLTWH(20, pageH - 20, pageWidth - 40, 12),
  );

  await _saveAndShare(document, 'Presupuesto_${event.clientName}.pdf');
  document.dispose();
}

// ─────────────────────────────────────────────
// CONTRATO PDF
// ─────────────────────────────────────────────
Future<void> generateContractPDF({
  required EventEntity event,
  required List<EventProductEntity> products,
  required List<EventExtraEntity> extras,
  String businessName = 'EventosApp',
}) async {
  final document = PdfDocument();
  final page = document.pages.add();
  final graphics = page.graphics;
  final pageWidth = page.getClientSize().width;
  final pageH = page.getClientSize().height;

  final brandColor = PdfColor(_brandR, _brandG, _brandB);
  final grayColor = PdfColor(102, 102, 102);
  final textColor = PdfColor(51, 51, 51);

  final boldFont =
      PdfStandardFont(PdfFontFamily.helvetica, 10, style: PdfFontStyle.bold);
  final regularFont = PdfStandardFont(PdfFontFamily.helvetica, 9);
  final titleFont =
      PdfStandardFont(PdfFontFamily.helvetica, 20, style: PdfFontStyle.bold);
  final headerFont =
      PdfStandardFont(PdfFontFamily.helvetica, 24, style: PdfFontStyle.bold);
  final bodyFont = PdfStandardFont(PdfFontFamily.helvetica, 9);

  double y = 20;

  // ── Encabezado ──
  graphics.drawString(
    businessName,
    titleFont,
    brush: PdfSolidBrush(brandColor),
    bounds: Rect.fromLTWH(20, y, pageWidth / 2, 25),
  );
  graphics.drawString(
    'CONTRATO',
    headerFont,
    brush: PdfSolidBrush(textColor),
    bounds: Rect.fromLTWH(0, y, pageWidth - 20, 25),
    format: PdfStringFormat(alignment: PdfTextAlignment.right),
  );
  y += 30;

  graphics.drawLine(
    PdfPen(brandColor, width: 1.5),
    Offset(20, y),
    Offset(pageWidth - 20, y),
  );
  y += 15;

  // ── Cuerpo del contrato ──
  final today = _formatDate(DateTime.now());
  final eventDateStr = _formatDate(event.eventDate);
  final city = event.city?.isNotEmpty == true ? event.city! : 'la ciudad';

  final body =
      '''En $city, a los $today, comparecen por una parte $businessName (en adelante "EL PROVEEDOR"), y por la otra parte ${event.clientName} (en adelante "EL CLIENTE"), quienes convienen en celebrar el presente Contrato de Prestación de Servicios.

PRIMERA: OBJETO.
EL PROVEEDOR se compromete a prestar los servicios de ${event.serviceType} para el evento que se llevará a cabo el día $eventDateStr.

SEGUNDA: DETALLES DEL EVENTO.
El evento contará con aproximadamente ${event.numPeople} personas.
Ubicación: ${event.location.isNotEmpty ? event.location : 'A definir'}.
Horario: ${event.startTime} - ${event.endTime}.

TERCERA: SERVICIOS CONTRATADOS.
${products.map((p) => '  • ${p.quantity.toStringAsFixed(0)}x ${p.productName}').join('\n')}${extras.isNotEmpty ? '\n\nExtras:\n${extras.map((e) => '  • ${e.description}').join('\n')}' : ''}

CUARTA: COSTO Y FORMA DE PAGO.
El costo total del servicio es de ${_formatCurrency(event.totalAmount)}.
EL CLIENTE realizará un anticipo del ${event.depositPercent.toStringAsFixed(0)}% (${_formatCurrency(event.depositAmount)}) para reservar la fecha, debiendo liquidar el saldo restante antes del inicio del evento.

QUINTA: CANCELACIONES.
En caso de cancelación por parte de EL CLIENTE con menos de ${event.cancellationDays.toStringAsFixed(0)} días de anticipación, el anticipo no será reembolsado.${event.refundPercent > 0 ? '\nSi la cancelación se realiza con la debida anticipación, se reembolsará el ${event.refundPercent.toStringAsFixed(0)}% del anticipo entregado.' : ''}

SEXTA: ACEPTACIÓN.
Ambas partes aceptan los términos y condiciones del presente contrato, firmando de conformidad.''';

  final textElement = PdfTextElement(
    text: body,
    font: bodyFont,
    brush: PdfSolidBrush(textColor),
    format: PdfStringFormat(lineSpacing: 4),
  );

  textElement.draw(
    page: page,
    bounds: Rect.fromLTWH(20, y, pageWidth - 40, pageH - y - 70),
  );

  // ── Firmas ──
  final signY = pageH - 55;
  graphics.drawLine(PdfPen(textColor), Offset(20, signY), Offset(90, signY));
  graphics.drawString('EL PROVEEDOR', regularFont,
      brush: PdfSolidBrush(grayColor),
      bounds: Rect.fromLTWH(20, signY + 4, 70, 10),
      format: PdfStringFormat(alignment: PdfTextAlignment.center));
  graphics.drawString(businessName, boldFont,
      brush: PdfSolidBrush(textColor),
      bounds: Rect.fromLTWH(20, signY + 16, 70, 10),
      format: PdfStringFormat(alignment: PdfTextAlignment.center));

  graphics.drawLine(PdfPen(textColor), Offset(pageWidth - 90, signY),
      Offset(pageWidth - 20, signY));
  graphics.drawString('EL CLIENTE', regularFont,
      brush: PdfSolidBrush(grayColor),
      bounds: Rect.fromLTWH(pageWidth - 90, signY + 4, 70, 10),
      format: PdfStringFormat(alignment: PdfTextAlignment.center));
  graphics.drawString(event.clientName, boldFont,
      brush: PdfSolidBrush(textColor),
      bounds: Rect.fromLTWH(pageWidth - 90, signY + 16, 70, 10),
      format: PdfStringFormat(alignment: PdfTextAlignment.center));

  await _saveAndShare(document, 'Contrato_${event.clientName}.pdf');
  document.dispose();
}

// ─────────────────────────────────────────────
// Helper: guardar y compartir
// ─────────────────────────────────────────────
Future<void> _saveAndShare(PdfDocument document, String filename) async {
  final bytes = await document.save();
  final dir = await getApplicationDocumentsDirectory();
  final file = File('${dir.path}/$filename');
  await file.writeAsBytes(bytes);
  await Share.shareXFiles([XFile(file.path)], text: filename);
}
